### PR TITLE
Add pcap file converter

### PR DESCRIPTION
**Pull request by fdanis-oss**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

File: contrib/pcap2emulation.py [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+
+if __name__ == "__main__":
+    if {"-?", "-h", "--help", "--usage"}.intersection(set(sys.argv)):
+        usage(0)
+
+    if len(sys.argv) < 4:
+        usage(1)
+
+    path = os.path.abspath(os.path.expanduser(os.path.expandvars(sys.argv[1])))
+    parser = Pcap2Emulation(sys.argv[2], sys.argv[3:])
```
Comment (superm1) [inline contrib/pcap2emulation.py:360]: why not use `argparse`?  It's a lot more flexible to expand arguments later and re-order and such.

File: docs/device-emulation.md [hunk 2]
```diff
@@ -69,3 +69,33 @@ For example:
     Decompressing…           [***************************************]
     Waiting…                 [***************************************]
     Hughski ColorHug2: OK!
+
+## Pcap file conversion
+
+Emulation can also be used during the development phase of the plugin if the hardware is not
+available or to reduce the testing time. But this requires a way to create the emulation file while
```
Comment (superm1) [inline docs/device-emulation.md:76]: probably also helps with the fact you don't need to do a write cycle with the device while writing the code too.   Nice!

File: contrib/pcap2emulation.py [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+                                    self._save_event(event_bytes)
+
+                            if descriptor_type == DESCRIPTOR_INTERFACE:
+                                interface = {
+                                    "Length": get_int(layers["usb_usb_bLength"][-1]),
+                                    "DescriptorType": 4,
+                                }
+                                val = get_int(layers["usb_usb_bInterfaceNumber"])
+                                if val != 0:
+                                    interface["InterfaceNumber"] = val
```
Comment (superm1) [inline contrib/pcap2emulation.py:259]: stuff like this seems to beg a dictionary matching table.    Something like this:
```
table = {
"usb_usb_bInterfaceNumber" : "InterfaceNumber"
}
for o in table:
    val = get_int(layers[o])
    if val != 0:
        interface[o[table]] = val
```

File: contrib/pcap2emulation.py [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+def get_int(str):
+    if str[:2] == "0x":
+        return int(str, 16)
+    return int(str)
+
+
+class Pcap2Emulation:
+    def __init__(self, prefix, device_ids):
+        self.device = None
+        self.prefix = prefix
```
Comment (hughsie) [inline contrib/pcap2emulation.py:35]: i'd feel a lot more warm and fuzzy if these had mypy annotations -- not a blocker by any means, but it's certainly saved me quite a few headaches when I wrote the LVFS...

**Comment by hughsie** [comment]
Can you squash these down into one commit please?

**Review by hughsie — APPROVED** [review]
This looks great!

**Comment by fdanis-oss** [comment]
@hughsie I'm not sure of the following change: https://github.com/fwupd/fwupd/compare/8e97a7c1c6f55098df39630b107dacd1801990d8..f0fdde5e0bdf58183f4082433552a513b19e4b7c

This works well for the emulation file generated for the Synaptics-Prometheus, allowing the firmware update to complete.

But for the emulation file generated for the CalDigit, even if the firmware update completes, this patch prevents the version update because the device is not re-enumerated during the reload phase. 

do you have any suggestion on how to detect the need for re-enumeration from the USB packets?

**Comment by fdanis-oss** [comment]
I updated the re-enumeration requirement detection by tracking the `C_PORT_CONNECTION` clear feature packet.

It allows me to complete firmware update with emulation files generated from `CalDigit`, `ColorHug` and `Synaptics-Prometheus` pcap files.

**Review by hughsie — APPROVED** [review]

**Comment by hughsie** [comment]
@fdanis-oss feel free to merge whenever you're happy.

**Pull request merged**